---
title: "Course work"
output:
  distill::distill_article:
    toc: true
    toc_depth: 2
---

## Summary

You will have three individual assignments, six seminar submissions, and one group project: 

* One intro assignment (5%)
* Seminar participation (30% total; after dropping the lowest score)
* One paper review (5%) 
* One analysis assignment (20%)
* Final group project (40%) which contains multiple parts. 

<font color="red">**Deadlines are all by 11:59 pm (Pacific time) on the due date. Any submission or modification after the due date will not be graded unless you have requested an extension**</font>. If you anticipate having trouble meeting a deadline and need to request an extension/academic concession please reach out via email.

For detailed instructions on how to submit your work, see the [submission guide](submission_guide.html). 

## Timeline Overview

For a visual summary, click [here](timeline.html).

```{r, echo=FALSE, warning=FALSE}
suppressPackageStartupMessages(library(tidyverse))
assn <- read.csv("course-admin/assn.csv")
assn <- assn %>% 
  select(Category, content, start) %>%
  filter(Category != "Seminar") %>% 
  rename(Assignment = content, `Due Date` = start) %>%
  mutate(`Due Date` = as.Date(`Due Date`, "%Y-%m-%d")) %>%
  arrange(`Due Date`) %>% 
  mutate(`Due Date` = format(`Due Date`, "%a %b %d"))
knitr::kable(assn)
```

Seminars (you have a week to submit your deliverables after each seminar):

```{r, echo=FALSE, warning=FALSE}
suppressPackageStartupMessages(library(tidyverse))
assn <- read.csv("course-admin/assn.csv")
assn <- assn %>% 
  select(Category, content, start) %>%
  filter(Category == "Seminar") %>% 
  rename(Assignment = content, `Date` = start) %>%
  mutate(`Date` = as.Date(`Date`, "%Y-%m-%d")) %>%
  arrange(`Date`) %>% 
  mutate(`Date` = format(`Date`, "%a %b %d"))
knitr::kable(assn)
```

# Assignments Overview

## Intro assignment (5%)

This assignment is designed to give you independent practice in the workflow used for completing and submitting course work: committing and pushing files to GitHub, formatting an R Markdown document, using R to do simple analyses, and writing about your results. Grade point values are listed in the assignment.

The instructions and questions are available [here](https://github.com/STAT540-UBC/intro-assignment/blob/main/intro_assignment.md).

## Seminar participation (30%)

You will submit short "deliverables" to demonstrate your participation for every seminar. These deliverables give practical experience applying the knowledge that will be helpful on homework, the project, and your future research. Each Seminar session is weighted equally, but the lowest score will be dropped. Seminar deliverables are due on the next Tuesday following the TA-led session for that seminar. See the [Seminars page](seminars.html) for the submission materials and schedule.

## Paper critique (5%)

Each student will review and provide a written critique (max 1000 words) of a paper that will be posted on Canvas.

Please see [**critique rubric**](critique.html) for detailed instructions on this task. 

## Analysis assignment (20%)

This assignment will assess your understanding of the seminar and lecture materials. Start early because this assignment will take time to fully work through. Use the issues in the Discussion repo and the seminar time to ask questions. You will find most of the analysis workflow of the assignment in the seminar materials. 

The instruction and questions will be available [here](https://github.com/STAT540-UBC/analysis-assignment/).

The grade point values for each question are listed right in the assignment.


## Final group project (40%)

### General principles

Identify a biological question of interest and a relevant dataset. Develop and apply a statistical approach that allows you to use the dataset to answer the question.

We assume the biological question and data fall in the general area of high-throughput, large-scale biological investigations *targeted by the course*. Beyond that, it is wide open: methylation, SNPs, miRNAs, CNVs, RNA-Seq, CHiP-Seq, gene networks, ... it's fair game. Avoid a dataset that doesn't have any/much quantitative data, i.e. contains only sequence or discrete data. If you are using published data, it is critical to *be clear about how your project differs from previous literature*.

Note that definitive answers are not necessarily expected. Rather, aim to provide a critical appraisal of the data, the analytical approach, and the results. You will have to handle the competing pressures to "get it right" and "get it done". Shortcomings of the data, misfits between the data or the biological question and the statistical model, etc. are inevitable. Your goal is to identify such issues and discuss them critically, without becoming paralyzed. Demonstrate understanding of the statistical concepts and methods that are the foundation of your analytical approach.

We assume the analytical and computing task will have a substantial statistical component, probably enacted via R. So beware of a major analytical or computational undertaking that is, nonetheless, not statistical (example: constructing a database). Creating useful data visualizations can be absolutely vital and is arguably statistical, but your analysis should go beyond merely creating pretty pictures (but please do include some!). Key concepts, at least some of which should come up in your analysis:

 * the (hypothesized, probably artificial) data-generating model

 * background variation, variance, signal to noise ratio, estimates and their associated standard error

 * relationship between biological factors and experimental factors, apparent relative importance in terms of "explaining" observed data

 * attention to large-scale inference, e.g. control of family-wise error rate or false discovery rate
 
 
### Data considerations 

#### Appropriate use of data

If your project involves using unpublished data, ensure your plans are known to the data providers (e.g., your supervisors), and think about implications for publishing - are you are bringing the project team in as collaborators in effect? Are you planning to publish the results of your project, and if so who will be the co-authors? It is best to deal with these questions at the outset of the project.

#### Privacy of project data

The projects are not made public (other than an oral presentation of your work in front of your classmates, which **will be recorded** and made available only to the teaching team and registered students in the course). The project report materials are loaded into GitHub, the secure site we use to manage the course. The course staff and instructors are the only people who have access to the project GitHub repo other than the other members of the project group. The data used can be uploaded to the project, but this can limited or omitted if there are special concerns about privacy etc. - it's primarily the code and write-up about the results that needs to be provided for evaluation.

You can read Github's [security](https://help.github.com/articles/github-security/) and [privacy](https://help.github.com/articles/github-privacy-statement/) policies.

### Group makeup

Groups will be formed by the instructional team following the results of an in class survey, and posted in the Discussion repo. Groups will have a target size of 4 members. Groups will be formed with priority for diversity in terms of backgrounds. In practice, this probably means the team members come from a mix of programs/departments. We will try to honour requests for working with specific team mates, and you may come talk to us immediately after receiving group assignments if you'd like to make a change.

### Deliverables

Details and grading rubrics for each component of the final group project can be found on the [**Group project rubrics page**](group_project_rubrics.html).

* [Initial Proposal (5%)](group_project_rubrics.html#initial-proposal-5-pts)
<!-- * [Final Proposal (5%)](group_project_rubrics.html#final-proposal-5-pts) -->
* [Progress report (5%)](group_project_rubrics.html#progress-report-8-pts)
* [Final report (10%)](group_project_rubrics.html#final-report-10-pts)
* [Oral presentation (10%)](group_project_rubrics.html#oral-presentation-15-pts) - A Detailed rubric can be found [here](course-admin/ProjectPresentationRubric.pdf)
* [Individual report (10%)](group_project_rubrics.html#individual-report-7-pts)
