---
title: "Lectures"
---


```{r include = FALSE}
library(tidyverse)
library(stringr)
library(httr)
```



```{r include = FALSE}
# Import schedule
rawdat <- read.csv(file.path("course-admin", "schedule.csv"), stringsAsFactors = FALSE, colClasses = list(course = 'factor', dow = 'factor'))
rawdat$course <- as.factor(rawdat$course)

# Class meetings
rawdat_cm <- rawdat[rawdat$course == "STAT540",]

cdat <- rawdat_cm %>%
  mutate(ugly_date = ISOdate(2026, match(month, month.name), day),
         date = sprintf("%s %s %s",
                        dow, months(ugly_date, abbreviate = TRUE),
                        day),
         notes = ifelse(grepl("^lect", lect) & raw_notes != "",
                        sprintf("%s: %s", lect, raw_notes), raw_notes))

## link to the class page of any class meetings where such exists
req <- GET("https://api.github.com/repos/STAT540-UBC/lectures/git/trees/main?recursive=1")
stop_for_status(req)
filelist <- unlist(lapply(content(req)$tree, "[", "path"), use.names = F)

# now grab pdf & html file names
cmdat <- 
  data.frame(file = filelist[grepl('^lect.*pdf|^lect.*html', basename(filelist))],
             stringsAsFactors = FALSE) %>% 
  mutate(lect = str_extract(file, "lect[0-9]+")) %>%
  mutate(lect = sapply(lect, str_replace, pattern="lect0*([0-9]+)",
                       replacement="lecture-\\1")) %>%
  mutate(type = ifelse(grepl("pdf", file), "pdf", "html")) %>%
  pivot_wider(id_cols = lect, names_from = type, values_from = file)

if (nrow(cmdat) >0){
  cdat <- cdat %>% left_join(cmdat, by="lect")
}else{
  cdat$file <- NA
}

pdf_exists <- grepl("*.pdf", cdat$pdf)
html_exists <- grepl("*.html", cdat$html)

if (length(html_exists) == 0){
  if (length(pdf_exists) > 0){
    html_exists <- rep(FALSE, length(pdf_exists))
  }
}

if (length(pdf_exists) == 0){
  if (length(html_exists) > 0){
    pdf_exists <- rep(FALSE, length(html_exists))
  }
}

## Insert placeholder link to notes 
##cdat$notes[pdf_exists] <-
## str_replace(cdat$notes[which(pdf_exists)],
##              "(lecture-[0-9]+)|(lecture-[1-9]+)", '<a href="LINK">\\1</a>')
cdat$notes <- gsub("lecture-", "Lecture ", cdat$notes)

## URL base constants
HTML_BASE_URL <- "https://stat540-ubc.github.io/lectures/"
PDF_BASE_URL <- "https://github.com/STAT540-UBC/lectures/raw/main/"

## Helper function to create HTML link
create_html_link <- function(html_file) {
    paste0(HTML_BASE_URL, html_file)
}

## Helper function to create single PDF URL
create_pdf_url <- function(pdf_file) {
    paste0(PDF_BASE_URL, pdf_file)
}

## Helper function to create PDF markdown links
create_pdf_links <- function(pdf_files, start_index = 1) {
    if (length(pdf_files) < start_index) return(NULL)

    indices <- start_index:length(pdf_files)
    pdf_links <- sapply(indices, function(j) {
        label <- if (length(pdf_files) == 1) "PDF" else paste0("PDF", j)
        paste0("[", label, "](", create_pdf_url(pdf_files[j]), ")")
    })
    paste(pdf_links, collapse = " ")
}

## Main function to generate note with links based on available files
generate_note_with_links <- function(note, html_file, pdf_files, has_html, has_pdf) {
    if (!has_html && !has_pdf) return(note)

    ## Determine primary URL and additional links
    if (has_html) {
        primary_url <- create_html_link(html_file)
        additional <- if (has_pdf) create_pdf_links(pdf_files) else NULL
    } else {
        primary_url <- create_pdf_url(pdf_files[1])
        additional <- create_pdf_links(pdf_files, start_index = 2)
    }

    ## Create note with primary link
    note_with_link <- str_replace(note, "(Lecture [0-9]+)", paste0('[\\1](', primary_url, ')'))

    ## Append additional links if any
    if (!is.null(additional)) note_with_link <- paste0(note_with_link, " ", additional)

    return(note_with_link)
}

## Apply function to generate notes with links
for (i in seq_along(cdat$notes)) {
    cdat$notes[i] <- generate_note_with_links(
        cdat$notes[i],
        cdat$html[i],
        cdat$pdf[[i]],
        html_exists[i],
        pdf_exists[i]
    )
}

dat_cm <- cdat %>% split(cdat$course)
dat_cm <- lapply(dat_cm, 'rownames<-', NULL)
```



### Lecture schedule 

**Time** : Tues Thurs 9:00 - 10:30am

**Location**: Building and room number listed on Canvas

```{r results = 'asis', echo = FALSE}
# add readings
sched <- dat_cm[["STAT540"]] %>%
  mutate(readings = "") %>%
  select(date, readings, notes, instructor)


## lect 1
sched$readings[1] <- "[1](https://stat540-ubc.github.io/syllabus.html)"

## lect 2
sched$readings[2] <- "[1](https://github.com/STAT540-UBC/lectures/blob/main/lect02-intro-stat-matrix/README.md)"

# lect 3
sched$readings[3] <- "[1](https://github.com/STAT540-UBC/lectures/blob/main/lect03-defining-problems/README.md)" 
  
# # lect 4
# sched$readings[4] <- "
# [1](http://genomicsclass.github.io/book/pages/populations_and_samples.html), 
# [2](http://genomicsclass.github.io/book/pages/random_variables.html), 
# [3](http://genomicsclass.github.io/book/pages/clt_and_t-distribution.html)
# "
# 
# lect hypothesis
sched$readings[4] <- "[1](https://github.com/STAT540-UBC/lectures/tree/main/lect04-hypothesis-testing/README.md)"

# lect RNA
sched$readings[5] <- "[1](https://github.com/STAT540-UBC/lectures/tree/main/lect05-rna-seq/README.md)"

# lect RNA
sched$readings[6] <- "[1](https://github.com/STAT540-UBC/lectures/tree/main/lect06-rna-seq-modelling/README.md)"

# # lect 6
# sched$readings[6] <- "
# [1](http://genomicsclass.github.io/book/pages/matrix_notation.html), 
# [2](http://genomicsclass.github.io/book/pages/matrix_operations.html), 
# [3](http://genomicsclass.github.io/book/pages/expressing_design_formula.html)
# "
# 
# # lect 7
# sched$readings[7] <- "[1](http://genomicsclass.github.io/book/pages/interactions_and_contrasts.html)"
# 
# lect 8
sched$readings[9] <- "[1](https://github.com/STAT540-UBC/lectures/blob/main/lect08-single-cell-technology/README.md)"

# lect 9
sched$readings[10] <- "[1](https://github.com/STAT540-UBC/lectures/blob/main/lect09-single-cell-factorization/README.md)"

# sched$readings[10] <- "[1](https://github.com/STAT540-UBC/lectures/blob/main/lect09-multipleTesting/README.md)"
# 
# lect 10
sched$readings[11] <- "[1](https://github.com/STAT540-UBC/lectures/blob/main/lect10-gwas/README.md)"
 
# # lect 11
# sched$readings[12] <- "[1](https://github.com/STAT540-UBC/lectures/blob/main/lect11-epigenetics/README.md)"
# 
# # lect 12
# sched$readings[15] <- "[1](https://github.com/STAT540-UBC/lectures/blob/main/lect12-gsea/README.md)"
# 
# # in-class project meetings
# sched$notes[8] <- "[Project proposal meetings](https://canvas.ubc.ca/courses/158233/pages/initial-proposal-feedback-meetings-thurs-jan-30-in-class)"

# print final table
knitr::kable(sched %>% 
  dplyr::rename(Date=date, "Pre-reading"=readings, Topic=notes, Instructor=instructor),
  "html", escape = FALSE) %>% 
  kableExtra::column_spec(1, width = "6em") %>%
  kableExtra::kable_styling(full_width = TRUE)
```

**Supplemental Files**: 

* Additional companion notes can be found in the [STAT540-UBC/resources GitHub repo](https://github.com/STAT540-UBC/resources)
* Source files for lectures can be found the [STAT540-UBC/lectures GitHub repo](https://github.com/STAT540-UBC/lectures)

